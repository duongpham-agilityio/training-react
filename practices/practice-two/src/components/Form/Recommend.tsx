import { memo, useCallback } from 'react';

// Styles
import styles from 'components/Form/index.module.css';

// Types
import { Recommend } from 'types';

export interface RecommendProps {
  books: Recommend[];
  onSelect: (index: number) => void;
}

const BookRecommend = ({ books, onSelect }: RecommendProps) => {
  const renderBook = useCallback(
    (book: Recommend, index: number) => {
      const { name } = book;
      const handleSelectBook = () => onSelect(index);

      return (
        <li
          className={styles.recommendItem}
          key={index}
          onClick={handleSelectBook}
        >
          <p>{name}</p>
        </li>
      );
    },
    [onSelect]
  );

  return <ul className={styles.recommend}>{books.map(renderBook)}</ul>;
};
export default memo(BookRecommend);
