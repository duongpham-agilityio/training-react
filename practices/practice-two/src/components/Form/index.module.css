.overlay {
  position: fixed;
  inset: 0;
  z-index: 9999;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: rgba(0, 0, 0, 0.2);
  font-family: var(--primary-font);
}

.form {
  width: 800px;
  height: 624px;
  background-color: white;
  border-radius: 15px;
  padding: 22px;
}

.heading {
  text-align: center;
  padding: 0 0 20px;
}

.content {
  display: flex;
  flex-direction: column;
  gap: 24px;
  padding: 20px 0;
}

.formItem {
  display: flex;
  align-items: start;
  gap: 60px;
}

.formItem .upload {
  position: relative;
  display: flex;
  align-items: start;
  gap: 26px;
}

.uploadItem:first-child {
  flex: 1;
}

.formField {
  flex: 1;
  position: relative;
}

.recommend {
  position: absolute;
  top: calc(100% + 20px);
  border-radius: 7px;
  background-color: white;
  z-index: 999;
  box-shadow: 0 6px 10px silver;
}

.recommendItem {
  padding: 10px 7px;
}

.recommendItem:hover {
  background-color: var(--gray-color);
  cursor: pointer;
}

.text {
  height: 118px;
}

.input {
  background-color: transparent;
  position: absolute;
}

.label {
  display: inline-block;
  font-weight: 700;
  padding: 10px 0;
}

.action {
  display: flex;
  align-items: center;
  justify-content: end;
  gap: 20px;
}

.image {
  width: 95px;
  height: 116px;
  object-fit: cover;
  border-radius: 7px;
}
