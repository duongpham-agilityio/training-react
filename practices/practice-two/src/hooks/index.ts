export * from './useBookForm';
export * from './useSearchParam';
export * from './usePagination';
export * from './useFilter';
export * from './useDebounce';
export * from './useBooks';
export * from './useBook';
export * from './usePopupContext';
export * from './useToastContext';
