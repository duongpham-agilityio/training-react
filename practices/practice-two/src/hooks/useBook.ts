import { useCallback } from 'react';
import { useParams } from 'react-router-dom';
import useSWR from 'swr';

// Types
import { Book } from 'types';

// Helpers
import { axiosConfig } from 'helpers';

// Constants
import { ENDPOINT } from '@constants';
import { usePopupContext } from './usePopupContext';

/**
 * Get the data of a book
 * @returns object containing properties and methods for interacting with a book
 */
export const useBook = () => {
  const { dispatch } = usePopupContext();
  const { id } = useParams();
  const { mutate, ...rest } = useSWR<Book>(`${ENDPOINT.BOOKS}/${id}`);

  const deleteBook = useCallback(() => {
    dispatch(() => {
      mutate(
        async () => {
          return await axiosConfig.delete(`${ENDPOINT.BOOKS}/${id}`);
        },
        {
          populateCache: false,
          revalidate: false,
        }
      );

      window.location.replace(`${ENDPOINT.BOOKS}`);
    });
  }, [dispatch, id, mutate]);

  return {
    ...{ ...rest, mutate },
    deleteBook,
  };
};
