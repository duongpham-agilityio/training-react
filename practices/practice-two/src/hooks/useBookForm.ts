/* eslint-disable no-irregular-whitespace */
/* eslint-disable @typescript-eslint/no-explicit-any */
import {
  ChangeEvent,
  FormEvent,
  useCallback,
  useMemo,
  useRef,
  useState,
} from 'react';

// Hook
import { useToastContext } from 'hooks';

// Helpers
import {
  axiosConfig,
  convertStringToTime,
  convertTimeToDate,
  validate,
} from 'helpers';

// Services
import { uploadImage } from 'services';

// Constants
import { MESSAGES, TIMEOUT_DEBOUNCE } from '@constants';

// Mock data
import { book } from 'mock-data';

// Types
import { Book, Recommend } from 'types';

export type FormData = Omit<
  Book,
  'publishDate' | 'deletedAt' | 'createdAt' | 'updatedAt'
> & {
  publishDate?: number;
};

/**
 * Interacting with forms
 * @param data initialization values â€‹of data fields in form
 * @param type To know if this is a create form or an update form
 * @returns Object
 */
export const useBookForm = (
  data: FormData,
  type: 'create' | 'update',
  submitHandler: (data: FormData) => Promise<void>
) => {
  const { setNotification } = useToastContext();
  const [state, setState] = useState<FormData & { imageName: string }>({
    ...data,
    imageName: '',
  });
  const [isUpload, setIsUpload] = useState(false);
  const [isSubmit, setIsSubmit] = useState(false);
  const [booksRecommended, setBooksRecommended] = useState<Recommend[]>([]);
  const refImage = useRef<HTMLInputElement>(null);
  const refTime = useRef<ReturnType<typeof setTimeout>>();
  /**
   * Convert data to render form
   */
  const value = useMemo(() => {
    const { publishDate, ...rest } = state;

    return {
      ...rest,
      publishDate: publishDate ? convertTimeToDate(publishDate) : '',
    };
  }, [state]);

  /**
   * Handling get recommend
   */
  const handleRecommended = useCallback((value: string) => {
    if (value) {
      if (refTime.current) clearTimeout(refTime.current);

      refTime.current = setTimeout(async () => {
        try {
          const books = await axiosConfig
            .get(`${process.env.VITE_RECOMMENDED_URL}?q=${value}`)
            .then((r) => r.data)
            .then((data) => {
              // eslint-disable-next-line @typescript-eslint/no-unused-vars
              const books =
                data?.items.map((_item: { volumeInfo: any }) => {
                  const { title, publishedDate, description, imageLinks } =
                    _item.volumeInfo;

                  return {
                    name: title,
                    publishDate: publishedDate,
                    description,
                    imageURL: imageLinks.thumbnail || book.imageURL,
                  };
                }) || [];

              return books;
            });

          setBooksRecommended(books);
        } catch (error) {
          setBooksRecommended([]);
        }
      }, TIMEOUT_DEBOUNCE);
    }
  }, []);

  /**
   * Handling select an recommend
   */
  const handleSelectRecommended = useCallback(
    (index: number) => {
      const book = booksRecommended[index];

      setState((prev) => ({
        ...prev,
        ...book,
      }));
      setBooksRecommended([]);
    },
    [booksRecommended]
  );

  const onSubmit = useCallback(
    (event: FormEvent) => {
      event.preventDefault();
      const { name, author, imageURL, description, publishDate } = state;
      const checked = {
        name,
        author,
        imageURL,
        description,
        publishDate,
      };
      const isError = validate(checked);

      if (isError) {
        return setNotification({
          message: MESSAGES.EMPTY_FIELD_DESCRIPTION,
          title: MESSAGES.EMPTY_FIELD,
          type: 'error',
        });
      }

      setIsSubmit(true);
      submitHandler(state).finally(() => setIsSubmit(false));
    },
    [setNotification, state, submitHandler]
  );

  const onChange = useCallback(
    async (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      const element = event.target;
      const value = element.value;
      const key: keyof typeof data = element.name as keyof typeof data;

      if (key === 'publishDate') {
        return setState((prev) => ({
          ...prev,
          publishDate: convertStringToTime(value.trim()),
        }));
      }

      if (key === 'imageURL') {
        const imageEl = element as HTMLInputElement;
        const files = imageEl.files;

        if (files) {
          const data = new FormData();
          const file = files[0];

          data.append('image', file);
          setIsUpload(true);
          uploadImage(data, (response) => {
            if (response.status) {
              const name = file.name;

              setIsUpload(false);
              setState((prev) => ({
                ...prev,
                imageURL: response.data.url,
                imageName: name,
              }));

              return;
            }

            return setNotification({
              message: MESSAGES.ERROR_TITLE,
              title: MESSAGES.EMPTY_FIELD,
              type: 'error',
            });
          });

          return;
        }

        return;
      }

      if (key === 'name') handleRecommended(value.trim());

      setState((prev) => ({
        ...prev,
        [key]: value,
      }));
    },
    [handleRecommended, setNotification]
  );

  return {
    value,
    isUpload,
    booksRecommended,
    refImage,
    isSubmit,
    handleSelectRecommended,
    resetRecommended: setBooksRecommended,
    onSubmit,
    onChange,
  };
};
