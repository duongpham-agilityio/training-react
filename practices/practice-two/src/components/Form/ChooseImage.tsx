import { memo } from 'react';

// Components
import { Button } from 'components/commons';

// Assets
import UploadIcon from 'assets/icons/upload.svg';

const ChooseImage = ({
  name,
  chooseHandler,
}: {
  name: string;
  chooseHandler: () => void;
}) =>
  !name ? (
    <Button
      label="Upload"
      width="w-lg"
      border="b-lg"
      variant="primary"
      leftIcon={UploadIcon}
      onClick={chooseHandler}
      data-testid="upload-btn"
    />
  ) : (
    <p onClick={chooseHandler}>{name}</p>
  );

export default memo(ChooseImage);
