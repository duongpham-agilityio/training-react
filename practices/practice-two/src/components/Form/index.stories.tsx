import { useCallback } from 'react';
import { Meta, Story } from '@storybook/react';

// Components
import Form, { FormProps } from 'components/Form';

// Hooks
import { FormData } from 'hooks';

export const Template: Story<FormProps> = () => {
  const onSubmit = useCallback(async (book: FormData) => {
    Promise.resolve(book);
  }, []);

  return (
    <Form
      value={{
        author: 'John',
        name: 'HTML/CSS',
        description: 'HTML/CSS with styles',
        imageURL: '',
        publishDate: new Date().getTime(),
      }}
      onSubmit={onSubmit}
    />
  );
};

export default {
  title: 'Practice/components/Form',
  component: Form,
} as Meta;
