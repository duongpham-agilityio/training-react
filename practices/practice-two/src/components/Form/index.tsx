import { MouseEvent, memo, useCallback } from 'react';
import isEqual from 'react-fast-compare';

// Hooks
import { useBookForm, FormData } from 'hooks';

// Components
import { Button, Heading, Input } from 'components/commons';
import BookRecommend from './Recommend';
import ChooseImage from './ChooseImage';

// Styles
import styles from 'components/Form/index.module.css';

export type FormProps = {
  value: FormData;
  title?: string;
  className?: string;
  type?: 'update' | 'create';
  onClose?: () => void;
  onSubmit: (data: FormData) => Promise<void>;
};

const Form = (props: FormProps) => {
  const {
    value: data,
    title = 'Create a new book marker',
    className = '',
    type = 'create',
    onClose,
    onSubmit,
  } = props;
  const {
    value: { author, description, imageURL, name, publishDate, imageName },
    refImage,
    isUpload,
    isSubmit,
    booksRecommended,
    handleSelectRecommended,
    resetRecommended,
    onChange,
    onSubmit: onSubmitOnHook,
  } = useBookForm(data, type, onSubmit);

  const chooseImage = useCallback(() => {
    refImage.current?.click();
  }, [refImage]);

  const resetRecommendedData = useCallback(
    (e: MouseEvent) => {
      e.stopPropagation();

      resetRecommended([]);
    },
    [resetRecommended]
  );

  return (
    <section className={styles.overlay} onClick={onClose}>
      <form
        className={`${styles.form} ${className}`}
        action="#"
        method="POST"
        onSubmit={onSubmitOnHook}
        onClick={resetRecommendedData}
        data-testid="book-form"
      >
        <Heading label={title} className={styles.heading} />

        <section className={styles.content}>
          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="bookName">
                Book name
              </label>
              <Input
                value={name}
                placeholder="Book name"
                name="name"
                id="bookName"
                onChange={onChange}
              />

              {!!booksRecommended.length && (
                <BookRecommend
                  books={booksRecommended}
                  onSelect={handleSelectRecommended}
                />
              )}
            </div>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="author">
                Author
              </label>
              <Input
                value={author}
                placeholder="Author"
                name="author"
                onChange={onChange}
              />
            </div>
          </div>

          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="publishedDate">
                Published date
              </label>
              <Input
                type="date"
                value={publishDate}
                name="publishDate"
                onChange={onChange}
              />
            </div>
            <div className={`${styles.formField} ${styles.upload}`}>
              <div className={styles.uploadItem}>
                <label className={styles.label} htmlFor="imageURL">
                  Upload
                </label>
                <Input
                  ref={refImage}
                  hidden
                  className={styles.input}
                  type="file"
                  name="imageURL"
                  id="imageURL"
                  accept=".png,.jpeg"
                  onChange={onChange}
                />

                <ChooseImage name={imageName} chooseHandler={chooseImage} />
              </div>
              <div className={styles.uploadItem}>
                {imageURL && (
                  <img
                    src={imageURL}
                    alt="image book"
                    className={styles.image}
                  />
                )}
              </div>
            </div>
          </div>

          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="bookName">
                Description
              </label>
              <Input
                variant="area"
                className={styles.text}
                value={description}
                placeholder="Book name"
                name="description"
                onChange={onChange}
              />
            </div>
          </div>

          <div className={styles.action}>
            <Button
              label="Cancel"
              width="w-lg"
              border="b-lg"
              onClick={onClose}
              disabled={isSubmit}
            />
            <Button
              disabled={isUpload || isSubmit}
              type="submit"
              label="Save"
              variant={isUpload ? 'default' : 'primary'}
              width="w-lg"
              border="b-lg"
            />
          </div>
        </section>
      </form>
    </section>
  );
};

export default memo(Form, isEqual);
