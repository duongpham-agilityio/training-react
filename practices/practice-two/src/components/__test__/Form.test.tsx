import { fireEvent, render } from '@testing-library/react';

// Hooks
import * as hooks from 'hooks';

// Mocks
import './mocks/image.test';

// Components
import Form, { FormProps } from 'components/Form';
import { RefObject } from 'react';

jest.mock('hooks');

const setup = (props: FormProps) => render(<Form {...props} />);

const onSubmit = jest.fn();
const mockProps: FormProps = {
  onSubmit,
  value: {
    author: 'dddd',
    description:
      'LIFE Magazine is the treasured photographic magazine that chronicled the 20th Century. It now lives on at LIFE.com, the largest, most amazing collection of professional photography on the internet. Users can browse, search and view photos of today’s people and events. They have free access to share, print and post images for personal use.',
    imageURL:
      'http://books.google.com/books/content?id=51YEAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api',
    name: 'LIFE23333',
    id: 16,
  },
};
const mockUseBookFormValue: ReturnType<typeof hooks.useBookForm> = {
  value: {
    author: 'dddd',
    description:
      'LIFE Magazine is the treasured photographic magazine that chronicled the 20th Century. It now lives on at LIFE.com, the largest, most amazing collection of professional photography on the internet. Users can browse, search and view photos of today’s people and events. They have free access to share, print and post images for personal use.',
    imageURL:
      'http://books.google.com/books/content?id=51YEAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api',
    name: 'LIFE23333',
    id: 16,
    publishDate: '1955-08-15',
    imageName: 'duong',
  },
  isUpload: false,
  isSubmit: false,
  booksRecommended: [],
  handleSelectRecommended: jest.fn(),
  resetRecommended: jest.fn(),
  onChange: jest.fn(),
  onSubmit: jest.fn(),
  refImage: {
    current: {
      click: null,
    },
  } as unknown as RefObject<HTMLInputElement>,
};

describe('Form component', () => {
  it('Match snapshot', () => {
    jest.spyOn(hooks, 'useBookForm').mockReturnValue(mockUseBookFormValue);

    const { container } = setup(mockProps);

    expect(container).toMatchSnapshot();
  });

  it('Render with props', () => {
    jest.spyOn(hooks, 'useBookForm').mockReturnValue({
      ...mockUseBookFormValue,
      isUpload: true,
      booksRecommended: [
        {
          description:
            'LIFE Magazine is the treasured photographic magazine that chronicled the 20th Century. It now lives on at LIFE.com, the largest, most amazing collection of professional photography on the internet. Users can browse, search and view photos of today’s people and events. They have free access to share, print and post images for personal use.',
          imageURL:
            'http://books.google.com/books/content?id=51YEAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api',
          name: 'LIFE23333',
        },
      ],
      value: {
        ...mockUseBookFormValue.value,
        imageName: '',
        imageURL: '',
      },
      refImage: {
        current: {
          click: jest.fn(),
        },
      } as unknown as RefObject<HTMLInputElement>,
    });
    const { getByTestId } = setup(mockProps);

    const formEl = getByTestId('book-form');

    fireEvent.click(formEl);

    expect(mockUseBookFormValue.resetRecommended).toBeCalled();
  });

  it('Choose image', () => {
    const chooseImage = jest.fn();

    jest.spyOn(hooks, 'useBookForm').mockReturnValue({
      ...mockUseBookFormValue,
      isUpload: false,
      booksRecommended: [],
      value: {
        ...mockUseBookFormValue.value,
        imageName: '',
      },
      refImage: {
        current: {
          click: chooseImage,
        },
      } as unknown as RefObject<HTMLInputElement>,
    });

    const { getByTestId } = setup(mockProps);
    const uploadBtn = getByTestId('upload-btn');

    fireEvent.click(uploadBtn);

    expect(chooseImage).toHaveBeenCalledTimes(1);
  });
});
