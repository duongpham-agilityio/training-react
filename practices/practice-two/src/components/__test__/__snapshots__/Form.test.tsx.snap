// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component Match snapshot 1`] = `
<div>
  <section>
    <form
      action="#"
      class="undefined "
      data-testid="book-form"
      method="POST"
    >
      <h2
        class="undefined undefined "
      >
        Create a new book marker
      </h2>
      <section>
        <div>
          <div>
            <label
              for="bookName"
            >
              Book name
            </label>
            <div
              class="undefined "
            >
              <button
                aria-label="Search button"
                class="undefined undefined
  undefined undefined
  undefined "
                style="background-color: transparent; padding: 0px;"
                type="button"
              />
              <input
                id="bookName"
                name="name"
                placeholder="Book name"
                type="text"
                value="LIFE23333"
              />
              <button
                aria-label="The right button"
                class="undefined undefined
  undefined undefined
  undefined "
                style="background-color: transparent; padding: 0px;"
                type="button"
              />
            </div>
          </div>
          <div>
            <label
              for="author"
            >
              Author
            </label>
            <div
              class="undefined "
            >
              <button
                aria-label="Search button"
                class="undefined undefined
  undefined undefined
  undefined "
                style="background-color: transparent; padding: 0px;"
                type="button"
              />
              <input
                name="author"
                placeholder="Author"
                type="text"
                value="dddd"
              />
              <button
                aria-label="The right button"
                class="undefined undefined
  undefined undefined
  undefined "
                style="background-color: transparent; padding: 0px;"
                type="button"
              />
            </div>
          </div>
        </div>
        <div>
          <div>
            <label
              for="publishedDate"
            >
              Published date
            </label>
            <div
              class="undefined "
            >
              <button
                aria-label="Search button"
                class="undefined undefined
  undefined undefined
  undefined "
                style="background-color: transparent; padding: 0px;"
                type="button"
              />
              <input
                name="publishDate"
                placeholder="Enter text..."
                type="date"
                value="1955-08-15"
              />
              <button
                aria-label="The right button"
                class="undefined undefined
  undefined undefined
  undefined "
                style="background-color: transparent; padding: 0px;"
                type="button"
              />
            </div>
          </div>
          <div
            class="undefined undefined"
          >
            <div>
              <label
                for="imageURL"
              >
                Upload
              </label>
              <div
                class="undefined "
              >
                <button
                  aria-label="Search button"
                  class="undefined undefined
  undefined undefined
  undefined "
                  style="background-color: transparent; padding: 0px;"
                  type="button"
                />
                <input
                  accept=".png,.jpeg"
                  hidden=""
                  id="imageURL"
                  name="imageURL"
                  placeholder="Enter text..."
                  type="file"
                />
                <button
                  aria-label="The right button"
                  class="undefined undefined
  undefined undefined
  undefined "
                  style="background-color: transparent; padding: 0px;"
                  type="button"
                />
              </div>
              <p>
                duong
              </p>
            </div>
            <div>
              <img
                alt="image book"
                src="http://books.google.com/books/content?id=51YEAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
              />
            </div>
          </div>
        </div>
        <div>
          <div>
            <label
              for="bookName"
            >
              Description
            </label>
            <textarea
              class="undefined "
              name="description"
              placeholder="Book name"
            >
              LIFE Magazine is the treasured photographic magazine that chronicled the 20th Century. It now lives on at LIFE.com, the largest, most amazing collection of professional photography on the internet. Users can browse, search and view photos of todayâ€™s people and events. They have free access to share, print and post images for personal use.
            </textarea>
          </div>
        </div>
        <div>
          <button
            class="undefined undefined
  undefined undefined
  undefined "
            type="button"
          >
            Cancel
          </button>
          <button
            class="undefined undefined
  undefined undefined
  undefined "
            type="submit"
          >
            Save
          </button>
        </div>
      </section>
    </form>
  </section>
</div>
`;
